### PR TITLE
[SYCL][ESIMD] Add histogram 256 bin test using SLM

### PR DESCRIPTION
**Pull request by cmc-rep**
Signed-off-by: Gang Y Chen <gang.y.chen@intel.com>

Sorry, I have to recreate this pull request with a new branch.  I am new to this. my old branch was in bad state
I have addressed all the comments in the old PR.

### CODE REVIEW

File: sycl/test/esimd/on-device/histogram_256_slm.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
+//==---------------- matrix_transpose.cpp  - DPC++ ESIMD on-device test ----==//
```
Comment (alexbatashev) [inline sycl/test/esimd/on-device/histogram_256_slm.cpp:1]:
```suggestion
//==--------------- histogram_256_slm.cpp  - DPC++ ESIMD on-device test ----==//
```

**Review by alexbatashev — APPROVED** [review]
Other comments from https://github.com/intel/llvm/pull/2387 can be considered resolved. LGTM.

File: sycl/test/esimd/on-device/histogram_256_slm.cpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+// TODO enable on Windows
+// REQUIRES: linux
+// REQUIRES: gpu
+// RUN: %clangxx-esimd -fsycl %s -o %t.out
+// RUN: %ESIMD_RUN_PLACEHOLDER %t.out
+
+#include "esimd_test_utils.hpp"
+
+#include <CL/sycl.hpp>
+#include <CL/sycl/intel/esimd.hpp>
```
Comment (alexbatashev) [inline sycl/test/esimd/on-device/histogram_256_slm.cpp:17]:
```suggestion
#include <CL/sycl/INTEL/esimd.hpp>
```

File: sycl/test/esimd/on-device/histogram_256_slm.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+  auto GlobalRange = cl::sycl::range<1>(num_threads);
+  auto LocalRange = cl::sycl::range<1>(NUM_BINS / 16);
+  cl::sycl::nd_range<1> Range(GlobalRange, LocalRange);
+
+  {
+    auto e = q.submit([&](cl::sycl::handler &cgh) {
+      cgh.parallel_for<class histogram_slm>(
+          Range, [=](cl::sycl::nd_item<1> ndi) SYCL_ESIMD_KERNEL {
+            histogram_atomic(input_ptr, output_surface, ndi.get_group(0),
+                             ndi.get_local_id(0), 16);
```
Comment (alexbatashev) [inline sycl/test/esimd/on-device/histogram_256_slm.cpp:166]: Does [get_global_linear_id()](https://intel.github.io/llvm-docs/doxygen/classcl_1_1sycl_1_1nd__item.html#a9d8f8ddfca1086f22464ff52f59502ee) behave incorrectly with ESIMD?

Comment (cmc-rep) [reply sycl/test/esimd/on-device/histogram_256_slm.cpp:166]: local-id is needed inside function for SLM access

File: sycl/test/esimd/on-device/histogram_256_slm.cpp (hunk 1 above)
Comment (cmc-rep) [reply sycl/test/esimd/on-device/histogram_256_slm.cpp]: updated PR

File: sycl/test/esimd/on-device/histogram_256_slm.cpp (hunk 2 above)
Comment (cmc-rep) [reply sycl/test/esimd/on-device/histogram_256_slm.cpp]: updated PR

**Review by alexbatashev — APPROVED** [review]

File: sycl/test/esimd/on-device/histogram_256_slm.cpp [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+#include <CL/sycl/INTEL/esimd.hpp>
+#include <iostream>
+
+static constexpr int NUM_BINS = 256;
+static constexpr int SLM_SIZE = (NUM_BINS * 4);
+static constexpr int BLOCK_WIDTH = 32;
+static constexpr int NUM_BLOCKS = 32;
+
+using namespace cl::sycl;
+using namespace sycl::intel::gpu;
```
Comment (bader) [inline sycl/test/esimd/on-device/histogram_256_slm.cpp:26]:
```suggestion
using namespace sycl::INTEL::gpu;
```

**Review by bader — CHANGES_REQUESTED** [review]

File: sycl/test/esimd/on-device/histogram_256_slm.cpp (hunk 4 above)
Comment (cmc-rep) [reply sycl/test/esimd/on-device/histogram_256_slm.cpp]: PR updated

**Review by alexbatashev — APPROVED** [review]

**Pull request merged**